### DISCUSSION TITLE
WindowsAppSDKSelfContained doesn't work

### DISCUSSION THREAD

**Discussion by khoabui1412** [microsoft/WindowsAppSDK#3150] [Q&A] [answered]
Hi,
I build an unpackage winUI3 app, and define in project's config:

>     <WindowsAppSDKSelfContained>True</WindowsAppSDKSelfContained>
>     <WindowsPackageType>None</WindowsPackageType>
My winUI3 app use WASDK version 1.1.5
However, when I copy and run .exe file on target device, it still requires to install Windows App Runtime, details:

> This application requires the Windows App Runtime Version 1.1
> (MSIX package version >= 1005.616.1651.0)
> Do you want to install a compatibile Windows App Runtime now?

I expect target device doesn't need to install any addtional runtime package, so how to do that?
Thanks.

**Comment by DarranRowe** [comment, marked as answer]
This is an issue with how the bootstrapper works.
The default for the auto initialisation, which happens when you use WindowsPackageType None with no other options, is to show the no match UI. However, it is dumb, it checks for the existence of the Windows App SDK but doesn't take self contained applications into account.
When you want to use the application self contained, you need to make sure that this option isn't used. The two options, [as documented](https://learn.microsoft.com/en-us/windows/apps/windows-app-sdk/use-windows-app-sdk-run-time), are to either use WindowsAppSDKBootstrapAutoInitializeOptions_OnNoMatch_ShowUI in your project file and set it to false:

```
<WindowsAppSDKBootstrapAutoInitializeOptions_OnNoMatch_ShowUI>false</WindowsAppSDKBootstrapAutoInitializeOptions_OnNoMatch_ShowUI>
```

or to manually call MddBootstrapInitialize2 yourself, in that case you wouldn't have WindowsPackageType in your project file. Having something like:

```
struct bootstrapper
{
	bootstrapper()
	{
		MddBootstrapInitialize2(WINDOWSAPPSDK_RELEASE_MAJORMINOR, WINDOWSAPPSDK_RELEASE_VERSION_TAG_W, {}, MddBootstrapInitializeOptions_None);
	}
	~bootstrapper()
	{
		MddBootstrapShutdown();
	}
};

static bootstrapper s_bootstrapper;
```

in one of your source files, or the C# equivalent, will initialize the Windows App SDK and you have direct control over the call.
